### PR TITLE
Rest Client Reactive: globally register providers annotated with @Provider

### LINKED ISSUE #18560: @RegisterProvider required for resteasy interfaces
## Description
When a filter is defined with @Provider is should be applied to the resteasy interface.

https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/quarkus-rest-client-reactive.20need.20to.20use.20.40RegisterProvider/near/245419083

### Expected behavior
The following code should be enough to apply a filter
@Provider  
public class LogClientResponseFilter implements ClientResponseFilter 
Currently the following must also be done:
It also has to be registered to the interface:
@RegisterProvider(LogClientResponseFilter.class)

### Actual behavior
The filter is not applied to the resteasy interface

## To Reproduce

[rest-client-reactive-quickstart.tar.gz](https://github.com/quarkusio/quarkus/files/6790488/rest-client-reactive-quickstart.tar.gz)

1. run test, check output
2. remove @RegisterProvider and verify that the output is missing

## Environment (please complete the following information):

uname -a
Linux rhclient444.xxx 3.10.0-1160.6.1.el7.x86_64 #1 SMP Wed Oct 21 13:44:38 EDT 2020 x86_64 x86_64 x86_64 GNU/Linux
java -version
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
Quarkus version: 1.11.1.Final
mvn -version
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Java version: 1.8.0_192, vendor: Oracle Corporation, runtime: /home/xxx/opt/jdk-8u192_b12/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "3.10.0-1160.6.1.el7.x86_64", arch: "amd64", family: "unix"

### PR DESCRIPTION
**Pull request by michalszynkiewicz**
fixes #18560

This change also adds support for sorting providers by `@Priority`

### CODE REVIEW

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java [hunk 1]
```diff
@@ -217,14 +213,34 @@ void registerProvidersFromAnnotations(CombinedIndexBuildItem indexBuildItem,
             constructor.invokeSpecialMethod(MethodDescriptor.ofConstructor(AnnotationRegisteredProviders.class),
                     constructor.getThis());
 
+            for (AnnotationInstance instance : index.getAnnotations(ResteasyReactiveDotNames.PROVIDER)) {
+                // TODO: we may want to filter out stuff that is server side only
```
Comment (michalszynkiewicz) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:217]: Here's the part I'm not really sure about.
@geoand do you think registering everything here can have some side effects?

Comment (gsmet) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:227]: You need to honor this: https://quarkus.io/guides/rest-client#rest-client-and-resteasy-interactions (or at least have an equivalent).

Comment (geoand) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:227]: Exactly. We already do the equivalent on the server side

Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:227]: I don't know how I could be unaware of this :) thank you! Fixing

**Review by geoand — APPROVED** [review]

**Comment by sberyozkin** [comment]
@michalszynkiewicz as we've discussed, 
`oidc-reactive-client-filter/OidcClientRecativeRequestFilter`  is meant to be registered selectively - you've confirmed that removing `@Provider` does not affect that the tests since the users are expected to use `@RegisterProvider`.

Note, with ResteasyClassic, `OidcClientRequestFilter`, marked with `@Provider` is registered either selectively with `@RegisterProvider` or globally for all the clients - it is important to be able to control it:

```
if (config.registerFilter) {
      // REGISTER for all the clients
      jaxrsProviders.produce(new ResteasyJaxrsProviderBuildItem(OidcClientRequestFilter.class.getName()));
} else {
      // selected registration is expected
      restAnnotationProvider.produce(new RestClientAnnotationProviderBuildItem(OIDC_CLIENT_FILTER,
                    OidcClientRequestFilter.class));
}


```

Note there `@Provider` has no side-effects. 

I'm concerned that just picking up all `@Provider` filters and applying them to all the clients globally may have side-effects. 
`@Provider` It is only meant to be a scanning hint - it was added to the spec before client filters were introduced. Besides MP Rest Client users are used to using `@RegisterProvider`. There will likely be duplicate registrations. 

IMHO it makes sense to address the issue of the duplicates as part of this PR, i.e if you have `@Provider` added  to `MyProvider` (I added it thinking it could be important for Quarkus to recognize it, like I said `OidcClientRequestFilter` has no problems in RestEasy Classic, yet) then it does not mean MP RestClient with `@RegisterProvider(MyProvider.class)` should have 2 instances of `MyProvider`.

**Comment by sberyozkin** [comment]
@michalszynkiewicz 

Please also check what the MP RestClient expectations around `@Provider` (since MP Rest Client != JAX-RS spec) are. if MP RestClient only knows about `RegisterProvider` then Quarkus also registering all applicable `@Provider` may indeed cause the duplicates. Registering `@provider` automatically for the client may be a good improvement - but there should be some guards to protects against the side-effects. For example, `@Feature` may also be registered in JAX-RS at either the client or server sides.
See the Feature docs, specifically:

```
 * <p>
 * Features implementing this interface MAY be annotated with the {@link javax.ws.rs.ext.Provider &#64;Provider}
 * annotation in order to be discovered by the runtime when scanning for resources and providers.
 * Please note that this will only work for server side features. Features in the Client API must
 * be registered programmatically.
 * </p>
```

See the note that at the client side only the programmatic registration is expected

**Comment by michalszynkiewicz** [comment]
thanks @sberyozkin !

MicroProfile Rest Client specification does not mention `@Provider` at all. I asked Andy, who leads the spec, how he handles it in his implementation.

Re the duplicates, what I currently have stores the providers to register (either registered `@RegisterProvider` or `@Provider`) in a map, so there will only be a single provider per class defined this way. 
Programmatically registered providers are a different story.

I can add a switch to disable automatic discovery for `@Provider`s for rest client, e.g. `quarkus.rest-client.provider-autodiscovery`, defaulting to `true`... I'll add it to this PR.

**Comment by michalszynkiewicz** [comment]
@geoand I added the switch, can you take another look if you still approve?

**Review by geoand — APPROVED** [review]

**Pull request merged**